### DISCUSSION TITLE
keyboard interrupt

### DISCUSSION THREAD

**Discussion by vihantyagi** [gaogaotiantian/viztracer#250] [Q&A] [answered]
Does viztracer fail to create the result file if the software to be profiled is killed via keyboard interrupt ? I do not get any output in this case. Command I'm using: viztracer -o result.html script.py

My workaround that worked was to break out of all the threads and exiting all processes after a certain time but it also includes too many changes in my code. Is there a way to profile the code using viztracer if the code is killed manually (keyboard interrupt or just kill -9 pid) after some time ? 

Thank you for your help

**Comment by gaogaotiantian** [comment, marked as answer]
It should save the result after keyboard interrupt.
Can you share your code to profile and the os you are working on? Also python and viztracer version.

**Reply by vihantyagi** [reply]
Ubuntu 20.04.4
Python 3.8.10
Viztracer 0.15.2
Unfortunately I cannot share my code with you. 

I have a simple sample script in which I am creating 2 processes and killing them manually using kill -9 pid. In this case I get a result.html file but I don't think it was profiled correctly (there is no separate timelines for the 2 processes and it shows that the main thread ran for only 150 ms which is incorrect). 

## sample script: 

class sample_class:

    def __init__(self):
        rtsp1 = <sample_rtsp>
        rtsp2 = <sample_rtsp>
        p = Process(target=self.hello, args=(rtsp1,))
        p1 = Process(target=self.test_rtsp, args=(rtsp2,))
        p.start()
        p1.start()

    def hello(self, rtsp):
        pid = os.getpid()
        while True:
            cap = cv2.VideoCapture(rtsp)
            ret, frame = cap.read()
            fps = cap.get(cv2.CAP_PROP_FPS)
            a = frame.shape
            image = cv2.line(frame, (40, 150), (100, 50), (0, 0, 255), 2)
            cv2.imwrite('img1.jpg', image)
            print("process1 ", pid, a, fps)
            print("process 1----------------------")

    def test_rtsp(self, rtsp):
        pid = os.getpid()
        time1 = time.time()
        while True:
            cap = cv2.VideoCapture(rtsp)
            ret, frame = cap.read()
            fps = cap.get(cv2.CAP_PROP_FPS)
            a = frame.shape
            image = cv2.line(frame, (40, 150), (100, 50), (0, 0, 255), 2)
            cv2.imwrite('img2.jpg', image)
            print("process 2",pid, a, fps)
            print('process 2----------------------')

if __name__ == "__main__":
    sample_object = sample_class()

## result:  
![image](https://user-images.githubusercontent.com/47202142/173249677-7fe1c3cc-8a27-4337-a125-6387ee3ac504.png)

Thank you for your help

**Reply by gaogaotiantian** [reply]
``kill -9``(or SIGKILL) won't leave the program any space to work, it will force the process to exit so it's impossible for VizTracer(or any other program) to do anything, that's expected. I'm more interested in the Keyboard Interrupt case (or SIGTERM(-15) related).

**Reply by vihantyagi** [reply]
Here's the traceback for keyboard interrupt for the sample_script:

```
^CError in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/usr/lib/python3.8/multiprocessing/popen_fork.py", line 27, in poll
    pid, sts = os.waitpid(self.pid, flag)
KeyboardInterrupt
Traceback (most recent call last):
  File "/root/Projects/yolo/bin/viztracer", line 8, in <module>
    sys.exit(main())
  File "/root/Projects/yolo/lib/python3.8/site-packages/viztracer/main.py", line 600, in main
    atexit._run_exitfuncs()
  File "/usr/lib/python3.8/multiprocessing/util.py", line 357, in _exit_function
    p.join()
  File "/usr/lib/python3.8/multiprocessing/process.py", line 149, in join
    res = self._popen.wait(timeout)
  File "/usr/lib/python3.8/multiprocessing/popen_fork.py", line 47, in wait
    return self.poll(os.WNOHANG if timeout == 0.0 else 0)
  File "/usr/lib/python3.8/multiprocessing/popen_fork.py", line 27, in poll
    pid, sts = os.waitpid(self.pid, flag)
KeyboardInterrupt

(yolo) root@a38b045c3e10:~/Raspberry-Pi-Object-Detection# Process Process-1:
Process Process-2:
Traceback (most recent call last):
Traceback (most recent call last):
  File "/usr/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/usr/lib/python3.8/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/usr/lib/python3.8/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/root/Raspberry-Pi-Object-Detection/sample_2.py", line 37, in test_rtsp
    cap = cv2.VideoCapture(rtsp)
  File "/root/Raspberry-Pi-Object-Detection/sample_2.py", line 25, in hello
    cap = cv2.VideoCapture(rtsp)
KeyboardInterrupt
KeyboardInterrupt
^C

```

**Reply by gaogaotiantian** [reply]
How many times did you click `Ctrl+C`?

**Reply by vihantyagi** [reply]
Twice. If I do it once it's just stuck there (at the last line in the above traceback).

**Reply by gaogaotiantian** [reply]
This is because your multi-process program can not be properly keyboard-interrupted. The first `Ctrl+C` raise the first KeyboardInterrupt and VizTracer is waiting for the subprocess to finish, but it can't. The second `Ctrl+C` stops the VizTracer so it won't be able to collect any information.

Without VizTracer, is your program be able to stop with a single `Ctrl+C`? VizTracer can't deal with a multi-process program that can't be properly stopped, because it requires some after-execution operations.

For a more formal program, you should try to achieve - the user can use Ctrl+C to exit the full program, but no exception is explicitly raised. The main process should catch that exception then properly terminate the subprocesses(by politely asking them or terminate them, not kill them).

**Reply by gaogaotiantian** [reply]
Can you try creating your `Process` with `daemon=True`?

**Reply by vihantyagi** [reply]
Yes you're right about the `Ctrl+C` issue (unable to kill with single `Ctrl+C` even without viztracer) you have described above. Setting the daemon = True worked for my sample script. I will update here if it works for the code I'm trying to profile after making relevant changes. 

Thank you so much for your help and quick replies !

**Comment by vihantyagi** [comment]
Also another I had one more question. Is log_parse compatible with multiprocess ?

**Reply by gaogaotiantian** [reply]
Yes, `log_sparse` works with multiprocess.
